docs/data-management.md

```diff
@@ -0,0 +1,72 @@
+# Data representation
+
+Icepeak is a key-value store for JSON data. In memory, the state is represented
+as a `Data.Aeson.Value`, on disk, it is serialized as the textual JSON
+representation.
+
+# Modifications
+
+Icepeak supports two kinds of modification operations:
+
+- `Put` writes a JSON value to a given path in the JSON structure.
+
+  - If the path descends into a JSON value that is not an object, that value is
+    replaced by an object that just contains the newly created path.
+  - If a value along the path is already an object, the field specified by the
+    path is added to it.
+
+- `Delete` removes a JSON value at a given path.
+
+  - If the path does not exist, nothing happens.
+  - If the path refers to a field in an object, that field is removed from the
+    object.
+
+# Atomic writes
+
+To make sure an interrupted write does not corrupt the database, the new value
+is first written to a file with the same name as the data file but with the
+extension `.new` appended. For example, if the data is stored in `icepeak.json`,
+the new data will first be written to `icepeak.json.new`.
+
+Only then, `icepeak.json.new` is renamed to `icepeak.json`, thereby overwriting
+it. POSIX guarantees that this rename operation is atomic.
+
+# Journaling
+
+Icepeak optionally supports journaling in order to prevent data loss.
```

Comment:
> means that the data has been written to disk

This is notoriously difficult to guarantee, especially in ~~a virtual machine running on unspecified hardware~~ the cloud. If strong durability is the goal, perhaps it would be worth using a backing store that already went through the trouble of trying to get this right.

Reply:
Isn't a file system such a backing store that tries to get that right?

Reply:
The best you can do is call fsync(2) and hope it gets actually written immediately.

Reply:
> In the future I would like to add an option to enable strong durability that guarantees that a successful HTTP response means that the data has been written to disk.

Radek had the idea that we could simply pass an `MVar` along with the command sent to the core, that is triggered as soon as the request has been processed.

We could also let the client decide whether to wait or not by adding a URL parameter `durable`.

Reply:
> We could also let the client decide whether to wait or not by adding a URL parameter durable.

That's a nice idea. It's more flexible than making this a service-wide setting.

Reply:
That should be quite easy to add, I could do it in this PR.

Reply:
> Isn't a (journaling) file system such a backing store that tries to get that right?

Not usually; btrfs for instance will by default only write to disk every 30 seconds. Even calling `fsync` on the file and directory [might not be sufficient](https://utcc.utoronto.ca/~cks/space/blog/unix/FileSyncProblem). And then even when the OS claims to have instructed the underlying device to write the data, that device could be a virtual disk, and the hypervisor might not write data to the physical device. And even when it does, the physical device or RAID controller will likely have a write cache that it will lose when power is lost. Expensive hardware has batteries that will last long enough to flush the write cache.